dbms/src/Flash/Mpp/MPPTunnel.cpp

```diff
@@ -104,6 +105,12 @@ void MPPTunnel::finishSendQueue()
 /// exit abnormally, such as being cancelled.
 void MPPTunnel::close(const String & reason)
 {
+    SCOPE_EXIT({
+        send_queue->finish();
+        MPPDataPacketPtr res;
+        while (send_queue->pop(res))
```

Comment:
Why do we need to pop the data explicitly, for the MPMCQueue dtor would do the same thing?

Reply:
Since MPMCQueue dtor will called after ProcListEntry dtor, the memory_tracker will have been already released. We need drain the MPMCQueue when `MPPTunenl::close` is called, since data in MPMCQueue is never used.

Reply:
added some explaination.

Reply:
But if he queue is cancelled, pop does not return the remaining object.

Reply:
fixed the cancel case